📖 Draft documentation for initial upgrade support (#559)
operator-framework/operator-controller: docs/drafts/initial-upgrade-support.md

# Initial upgrade support

This document explains how OLM 1.0 handles upgrades.

OLM 1.0 introduces initial upgrade support with simplified UX for package authors and package admins which allows defining upgrade edges implicitly via [Semantic Versioning](https://semver.org/) standard.

It also introduces an API to enable independently verified upgrades and downgrades.

## Upgrade constraint semantics

OLM 1.0 at the moment supports two upgrade constraint semantics:

* [Semantic Versioning](https://semver.org/) (Semver)
* [Legacy OLM 0 semantics](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#methods-for-specifying-updates)

**Note:** At the moment of writing legacy semantics support in OLM 1.0 is incomplete: only `replaces` directive is taken into account.

Kubernetes manifests provided in this repo enable Semver support by default, but cluster admins can control which semantics to use by passing one of the arguments into `manager` binary:
* `--feature-gates=ForceSemverUpgradeConstraints=true` - enable Semver
* `--feature-gates=ForceSemverUpgradeConstraints=false` - disable Semver, use legacy semantics

> 

Author: Added an example calls. Or should we include example manifests?

> 

Reviewer: I think including example manifests would be helpful. Because you can set the argument using either the CLI or by editing the manifest/yaml file?

And sorry for not mentioning it yesterday, but I am not familiar with the `manager` binary. Is that something new to the operator-controller or is it a third party tool? Do we need to introduce how to download the binary or include it as a prerequisite?

> 

Author: @michaelryanpeter the `manager` binary is operator-controller itself. This repo contains this manifest:

https://github.com/operator-framework/operator-controller/blob/6cb7029801558cd9e69fd4285468e600f45df7a6/config/manager/manager.yaml#L68-L72

But we later customise args like so:

https://github.com/operator-framework/operator-controller/blob/6cb7029801558cd9e69fd4285468e600f45df7a6/config/default/manager_auth_proxy_patch.yaml#L48-L53

Do you think we should have a full `Deployment` example for both states of the flag? Maybe under a spoiler? Or should we include only a part of the deployment manifest? E.g.

```yaml
- command:
  - /manager
  args:
  - --feature-gates=ForceSemverUpgradeConstraints=true
  image: controller:latest
```

> 

Reviewer: I think it would be work explaining. I read the instructions like I needed to enter a CLI command to enable the command, something like `kubectl manager --feature-gates=<etc>`.

I think that at this stage of development, just an excerpt like your last example would work. Then when we stitch the version, version range, channel, and upgrade docs together, we can maybe add another formulary of real world example manifests.

**Note:** There is currently no support for automatic upgrades to a next major version. For now such upgrades will have to be independently verified and manually preformed (more on this below).

### Upgrades within the major version zero

As per Semver major version zero is for initial development and anything may change at any time. Given that there are some special case for upgrades within the major version zero:

* It is not possible to automatically upgrade one patch versions to another when both major and minor versions are `0`. Meaning no automatic upgrades withing range `>= 0.0.1 <0.1.0`.
* It is not possible to automatically upgrade from one minor version to another minor version within the major version zero. E.g. no upgrades from `0.1.0` to `0.2.0`. But one can upgrade from patch to patch version. E.g. upgrades are possible in ranges `>= 0.1.0 <0.2.0`, `>= 0.2.0 <0.3.0`, `>= 0.3.0 <0.4.0` and so on.

To perform an upgrade in the cases mentioned above - the upgrade must be independently verified.

## Independently verified upgrades and downgrades

> **Warning:** This option requires thorough manual verification of the desired operation prior to applying to production workloads. Depending on the package, using this option might lead to catastrophic consequences such as data loss.

Reviewer: ```suggestion
**Warning:** If you want to force an upgrade manually, you must thoroughly verify the outcome before applying any changes to production workloads. Failure to test and verify the upgrade might lead to catastrophic consequences such as data loss.
```

> In cases when package admins need to upgrade or downgrade to a version which is not guaranteed to be compatible they can choose to set `.spec.upgradeConstraintPolicy` to `Ignore` on the relevant `Operator` resource.

Reviewer: ```suggestion
As a package admin, if you must upgrade or downgrade to version that might be incompatible with the currently installed version, you can set the `.spec.upgradeConstraintPolicy` field to `Ignore` on the relevant `Operator` resource.
```

> This option means that no upgrade constraints will be set on the package and version can be changed to any version available in catalogs for a given package.

Reviewer: ```suggestion
If you set the field to `Ignore`, no upgrade constraints are set on the package. As a result, you can change the version to any version available in the catalogs for a given package.
```
